File: src/main/java/live/adamlearns/animalroyale/Players.java
package live.adamlearns.animalroyale;

import java.util.HashMap;
import java.util.Map;

/**
 * This keeps track of everyone who has typed a command via Twitch.
 */
public class Players {
    private final Map<String, GamePlayer> players = new HashMap<>();

    public void removeAllPlayers() {
        players.clear();
    }

    public GamePlayer getPlayer(final String name) {
        return players.get(name);
    }

    public Map<String, GamePlayer> getAllPlayers() {
        return players;
    }

    public int getNumLivingSheep() {
        return (int) players.values().stream().filter(GamePlayer::isSheepAlive).count();
    }

    public GamePlayer getLastRemainingPlayer() {
        return players.values().stream().filter(GamePlayer::isSheepAlive).findFirst().orElse(null);
    }

    public GamePlayer createPlayerIfNotExists(final String name) {
        if (players.containsKey(name)) {
            return players.get(name);
        }

        final GamePlayer gamePlayer = new GamePlayer(name);
        players.put(name, gamePlayer);

        return gamePlayer;
    }
}

File: src/main/kotlin/live/adamlearns/animalroyale/Players.kt
package live.adamlearns.animalroyale

/**
 * This keeps track of everyone who has typed a command via Twitch.
 */
class Players {
    private val players: MutableMap<String, GamePlayer?> = HashMap()

    val allPlayers: Map<String, GamePlayer?>
        get() = players

    val numLivingSheep: Int
        get() = players.values.count { it?.isSheepAlive ?: false }

    val playersWithLivingSheep: List<GamePlayer?>
        get() = players.values.filter { it?.isSheepAlive ?: false }

    val lastRemainingPlayer: GamePlayer?
        get() = players.values.firstOrNull { it?.isSheepAlive ?: false }

    fun removeAllPlayers() {
        players.clear()
    }

    fun getPlayer(name: String): GamePlayer? {
        return players[name]
    }

    fun createPlayerIfNotExists(name: String): GamePlayer? {
        if (players.containsKey(name)) {
            return players[name]
        }

        val gamePlayer = GamePlayer(name)
        players[name] = gamePlayer

        return gamePlayer
    }
}
